V1.1 - Implemented modularity & improved code readability

Allyday opened the pull request:
## What?
I've applied OOP principles to improve code modularity. I also refactored most of the code using more human-readable names and less calculation codes with no comments. 

## Why?
These changes make code reviews easy, and future maintenance & enhancements even much more easier. While refactoring, I myself have gotten a few ideas on how the game can be enhanced in the future.

## How?
Most of my refactoring work was done while having OOP in mind.
- [x] Create `CanvasObject` and `Ball` classes to use as nodes instead of plain objects.
- [x] Use inheritance to create `Ball` class with extra capabilities (i.e bouncing).
- [x] Move collision detection methods inside `CanvasObject`.
- [x] Move movement methods inside `CanvasObject`.
- [x] Move bounce angle calculation inside `Ball` class.
- [x] Remove hardcoded calls to `app` object inside its own methods.
- [x] Move all calculations into specialized methods.
- [x] Give human-readable names to math-filled methods such as those related to collision detection, or bounce angle calculation.
- [x] Add comments, especially to math-filled methods.

## Screenshots
There is **no change to the gameplay**. I'm attaching a video from v1.0 for your reference.

https://user-images.githubusercontent.com/59161372/136654527-acab4f60-664f-4092-9436-3c09bf3a1ff3.mov

## Other - Ideas for 2.0
For the next version, we can consider these features:
- Bouncing from irregularly angled nodes (instead of perfectly vertical paddles).
- Text drawing onto canvas.
- Score tracking using text drawing.
- Difficulty settings: changing ball's & players' velocity.
- Refined styling, such as a round or disco-coloured ball.

Allyday committed:
create classes CanvasObject and Ball

Allyday committed:
change in-functions app calls to this (avoid hardcoding)

Allyday committed:
move collision & movement logics into nodes instead of app

demiculus commented:
@Allyday 

1. How many hours did this take (research & implementation)?
2. What's your full-time job? and how much are you paid gross vs net (not sure which is mentioned in Vietnam)?
3. What are your evening classes? What is your reason of taking them?
4. What queries did you research and read to implement these?
5. All these changes go into modularity which helps code be clean but I wouldn't necessarily say they are clean code principles. What is your take on it? 
6. Did you study uni? What did you study and where?

Allyday commented:
@demiculus 
> 1. How many hours did this take (research & implementation)?

It took me about 5 hours, which in hindsight is a shame since I could do more work on making clear commits.

> 2. What's your full-time job? and how much are you paid gross vs net (not sure which is mentioned in Vietnam)?

I'm currently a BA/Developer in a semi-govermental unit. We work in traceability, products info verification and more. I'm paid $500 gross, which is about ~$476 net monthly.
To note, I've handed in my notice and October 31st will be my last day.

> 3. What are your evening classes? What is your reason of taking them?

I'm studying computer programming at FPT Aptech Vietnam, which grants me an internationally recognized certificate on completion, and some universities take that as credits for 1-2 years. This is how I'm getting my bachelor degree.

> 4. What queries did you research and read to implement these?

My Google searches, chronologically:
```
modularity in javascript
interface javascript
inheritance in javascript
class javascript
include file js
cannot use import statement outside a module
cannot use import statement outside a module javascript
PR descriptions
pr descriptions github
```

> 5. All these changes go into modularity which helps code be clean but I wouldn't necessarily say they are clean code principles. What is your take on it?

With more (rather basic) research, here are the things that I think make my code not clean:
- Complex and non-descriptive calculations/checks, namely the ball's angle calculations and the whole collision prevention block.
- Inconsistent casing for constants.
- Unnecessary comments where code is actually descriptive.
- Too many arguments for constructor functions (I'm split on this one as it was used in production at my old job).
- The switchColors in my utilities has side effects, which shouldn't be the case for util functions.
- Not sure where this falls but I realized I implemented different collision prevention logics for the ball vs the players. This could have been one function, moved into CanvasObject class for future use & enhancement.

> 6. Did you study uni? What did you study and where?

As I mentioned above, I don't have a degree yet. I did go to uni but dropped out in my first year. I majored in Nanotechnology before dropping out. I worked for a couple years before deciding to pursue software development.

demiculus commented:
@Allyday 

> My Google searches, chronologically:

You should use `best practices` and `best principles`  keywords in your future queries

> As I mentioned above, I don't have a degree yet. I did go to uni but dropped out in my first year. I majored in Nanotechnology before dropping out. I worked for a couple years before deciding to pursue software development.

Why did you drop out? What was the decision making process like while you decide to pursue software development? 

Lets do an interview, please book a slot, let me know if the hours don't work and suggest alternatives: https://calendly.com/colonist-demi

demiculus commented:
> Lets do an interview, please book a slot, let me know if the hours don't work and suggest alternatives:

Can we do an hour earlier?

Allyday commented:
> Can we do an hour earlier?

Sure thing. I only see a 30-min earlier timeslot so I'm assuming I don't need to reschedule with Calendly?

Allyday commented:
@demiculus 
> You should use best practices and best principles keywords in your future queries

Thanks, just now, while doing more reading to answer your question, I have been made aware of how much more I can learn from those kinds of queries.

> Why did you drop out? What was the decision making process like while you decide to pursue software development?

While in uni, I was working part-time as a TA at small English center. It was my first job and it was also what made me realized I didn't only identify as an anti-social science nerd. I liked science but would have been bored to death if I were to stay in a lab forever. I like hanging out with people from different cultures, talking to them about my own culture. I also like travelling, and I plan on getting a good job to afford it :D

So I quit uni, and looked for a fulltime job. I got a job at a rather successful startup company that published their own English self-study books. It wasn't tech-related but I really enjoyed product development, working in a young team full of ideas & designing something that works. Something that is smart. I did know I wouldn't be there forever.

Programming was in my skills bucket list, and after a month learning it, I realized it's what I want to do for the long term. For me, it's significantly more satisfying than making a book. So I later quit my job to focus on self-studying, got an internship in frontend development and so I am here :)

demiculus commented:
> > Can we do an hour earlier?
> 
> Sure thing. I only see a 30-min earlier timeslot so I'm assuming I don't need to reschedule with Calendly?

I resheduled

demiculus commented:
> I did know I wouldn't be there forever.

How, why?

> Programming was in my skills bucket list

How did it end up there? What else is there?

Allyday commented:
> I did know I wouldn't be there forever.

It was something that I could do, not something that I wanted to do. It's cheesy to say but I didn't feel passionate doing the actual job, despite all other factors being quite okay.
I was good at Maths in school, which could be an explanation for why I enjoy programming much better? It's very similar to solving a hard problem.

> How did it end up there? What else is there?

Other things on my skills list:
- Photoshop
- Basic video editing
- Photography
- Playing an instrument, maybe guitar or ukulele
- Rollerblading/Iceskating
- Perspective drawing
- Data analysis
- ...The list continues

Some things, like programming or photoshop, are on the list because:
- It's learnable (at least compared to 3d design, for example).
- It makes a noticeable difference in my life if I learn it successfully.
- It's viewed as challenging so it would be nice to be able to do it.

Some others are just things I have tried, enjoyed, and thus would like to at least become someone okay at it, like iceskating or drawing.

Data analysis is recently added to the list, after  I came across a few articles about the field. To be able to draft up the right data to answer questions that users didn't even know they wanted to ask, is a skill I very much want to learn.
